File: ios/Hapi/Features/Chat/Blocks/QuestionToolPresentation.swift
import Foundation
import HapiClient
import HapiProtocol

/// Read-only projection, separate from drafts and permission submission. The
/// recorded input/result/answers remain untouched and available under Source.
struct QuestionToolDetails: Equatable, Sendable {
    let questions: [QuestionDetail]
    let showResult: Bool
    var hasAnswers: Bool { questions.contains { $0.hasAnswers } }
}

struct QuestionDetail: Equatable, Sendable {
    let header: String?
    let question: String
    let multiple: Bool
    let options: [QuestionOptionDetail]
    let otherAnswers: [String]
    let note: String?
    var hasAnswers: Bool { options.contains { $0.selected } || !otherAnswers.isEmpty || note != nil }
}

struct QuestionOptionDetail: Equatable, Sendable {
    let label: String
    let description: String?
    let selected: Bool
}

func isQuestionDetailsTool(_ name: String) -> Bool {
    let name = toolPresentationName(name)
    return isAskUserQuestionToolName(name) || isRequestUserInputToolName(name)
}

func questionToolDetails(_ tool: ChatToolCall) -> QuestionToolDetails {
    let name = toolPresentationName(tool.name)
    let answers = questionAnswerMap(tool.permission?.answers)
    let questions: [QuestionDetail]
    if isAskUserQuestionToolName(name) {
        let cursor = isCursorAskQuestionToolName(name)
        let parsed = parseAskUserQuestions(tool.input, cursorDialect: cursor)
        if parsed.isEmpty {
            // The existing malformed-input answer form submits free text as "0".
            let values = (answers["0"] ?? []).map(trimQuestionAnswer).filter { !$0.isEmpty }
            questions = values.isEmpty ? [] : [QuestionDetail(
                header: nil, question: "", multiple: false, options: [], otherAnswers: values, note: nil
            )]
        } else {
            questions = parsed.enumerated().map { index, question in
                let values = (answers[question.answerKey(index: index, useStableIds: cursor)] ?? [])
                    .map(trimQuestionAnswer).filter { !$0.isEmpty }
                let optionValues = question.options.map { cursor ? ($0.id ?? $0.label) : $0.label }
                return QuestionDetail(
                    header: question.header, question: question.question, multiple: question.multiSelect,
                    options: question.options.enumerated().map { index, option in
                        QuestionOptionDetail(label: option.label, description: option.description,
                                             selected: values.contains(optionValues[index]))
                    },
                    otherAnswers: values.filter { !optionValues.contains($0) }, note: nil
                )
            }
        }
    } else if isRequestUserInputToolName(name) {
        let parsed = parseRequestUserInputQuestions(tool.input)
        let live = requestQuestionDetails(parsed, answers: answers)
        // Prefer usable live answers as a whole; do not merge with stale history.
        if live.contains(where: { $0.hasAnswers }) {
            questions = live
        } else {
            var result = tool.result
            if case .string(let text) = result {
                result = try? JSONDecoder().decode(JSONValue.self, from: Data(text.utf8))
            }
            questions = requestQuestionDetails(parsed, answers: questionAnswerMap(result))
        }
    } else {
        questions = []
    }
    return QuestionToolDetails(questions: questions,
                               showResult: tool.state == .error || !questions.contains { $0.hasAnswers })
}

private func trimQuestionAnswer(_ value: String) -> String {
    value.trimmingCharacters(in: .whitespacesAndNewlines)
}

/// Only the documented flat/nested maps and one outer `answers` wrapper.
/// Try a direct map first: a question may itself be named "answers".
private func questionAnswerMap(_ value: JSONValue?) -> [String: [String]] {
    guard let object = value?.chatObject else { return [:] }
    func parse(_ object: [String: JSONValue]) -> [String: [String]] {
        object.compactMapValues { value in
            (value.chatArray ?? value.chatObject?["answers"]?.chatArray)?.compactMap(\.chatString)
        }
    }
    let direct = parse(object)
    if !direct.isEmpty { return direct }
    return object["answers"]?.chatObject.map(parse) ?? [:]
}

private func requestQuestionDetails(
    _ questions: [RequestUserInputQuestion], answers: [String: [String]]
) -> [QuestionDetail] {
    questions.map { question in
        var selected: Set<String> = []
        var other: [String] = []
        var note: String?
        for value in answers[question.id] ?? [] {
            let trimmed = trimQuestionAnswer(value)
            // An actual option named "user_note: ..." is not metadata.
            if let option = question.options.first(where: { $0.label == trimmed }) {
                selected.insert(option.label)
            } else if value.hasPrefix("user_note: ") {
                let text = String(value.dropFirst("user_note: ".count))
                if question.inputType == "editor" {
                    note = text // An intentionally empty document is still an answer.
                } else {
                    let text = trimQuestionAnswer(text)
                    if !text.isEmpty { note = text }
                }
            } else if !trimmed.isEmpty {
                // Codex history may contain plain free text without a note prefix.
                other.append(trimmed)
            }
        }
        return QuestionDetail(
            header: nil, question: question.question, multiple: question.multiple,
            options: question.options.map {
                QuestionOptionDetail(label: $0.label, description: $0.description, selected: selected.contains($0.label))
            },
            otherAnswers: other, note: note
        )
    }
}

File: android/app/src/main/kotlin/app/hapi/companion/feature/chat/blocks/QuestionToolPresentation.kt
package app.hapi.companion.feature.chat.blocks

import app.hapi.companion.feature.chat.permissions.RequestUserInputQuestion
import app.hapi.companion.feature.chat.permissions.isCursorAskQuestionToolName
import app.hapi.companion.feature.chat.permissions.parseAskUserQuestions
import app.hapi.companion.feature.chat.permissions.parseRequestUserInputQuestions
import app.hapi.protocol.chat.ChatToolCall
import app.hapi.protocol.chat.isAskUserQuestionToolName
import app.hapi.protocol.chat.isRequestUserInputToolName
import app.hapi.protocol.wire.HapiJson
import app.hapi.protocol.wire.arrayOrNull
import app.hapi.protocol.wire.objOrNull
import app.hapi.protocol.wire.stringOrNull
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/** Read-only projection, independent of answer drafts and permission submission. */
internal data class QuestionToolDetails(val questions: List<QuestionDetail>, val showResult: Boolean) {
    val hasAnswers: Boolean get() = questions.any { it.hasAnswers }
}

internal data class QuestionDetail(
    val header: String?,
    val question: String,
    val multiple: Boolean,
    val options: List<QuestionOptionDetail>,
    val otherAnswers: List<String>,
    val note: String?,
) {
    val hasAnswers: Boolean get() = options.any { it.selected } || otherAnswers.isNotEmpty() || note != null
}

internal data class QuestionOptionDetail(val label: String, val description: String?, val selected: Boolean)

internal fun isQuestionDetailsTool(name: String): Boolean {
    val name = toolPresentationName(name)
    return isAskUserQuestionToolName(name) || isRequestUserInputToolName(name)
}

internal fun questionToolDetails(tool: ChatToolCall): QuestionToolDetails {
    val name = toolPresentationName(tool.name)
    val answers = questionAnswerMap(tool.permission?.answers)
    val questions = when {
        isAskUserQuestionToolName(name) -> {
            val cursor = isCursorAskQuestionToolName(name)
            val parsed = parseAskUserQuestions(tool.input, cursorDialect = cursor)
            if (parsed.isEmpty()) {
                // The malformed-input answer form submits free text as "0".
                val values = answers["0"].orEmpty().map(String::trim).filter(String::isNotEmpty)
                if (values.isEmpty()) emptyList() else listOf(
                    QuestionDetail(null, "", false, emptyList(), values, null),
                )
            } else parsed.mapIndexed { index, question ->
                val values = answers[question.answerKey(index, useStableIds = cursor)].orEmpty()
                    .map(String::trim).filter(String::isNotEmpty)
                val optionValues = question.options.map { if (cursor) it.id ?: it.label else it.label }
                QuestionDetail(
                    header = question.header, question = question.question, multiple = question.multiSelect,
                    options = question.options.mapIndexed { optionIndex, option ->
                        QuestionOptionDetail(option.label, option.description, optionValues[optionIndex] in values)
                    },
                    otherAnswers = values.filter { it !in optionValues }, note = null,
                )
            }
        }
        isRequestUserInputToolName(name) -> {
            val parsed = parseRequestUserInputQuestions(tool.input)
            val live = requestQuestionDetails(parsed, answers)
            // Prefer usable live answers as a whole; never merge stale history.
            if (live.any { it.hasAnswers }) live else {
                val result = tool.result?.stringOrNull?.let { runCatching { HapiJson.parseToJsonElement(it) }.getOrNull() }
                    ?: tool.result
                requestQuestionDetails(parsed, questionAnswerMap(result))
            }
        }
        else -> emptyList()
    }
    return QuestionToolDetails(questions, showResult = tool.state == "error" || questions.none { it.hasAnswers })
}

/** Documented flat/nested maps and one outer wrapper; "answers" is also a valid question id. */
private fun questionAnswerMap(value: JsonElement?): Map<String, List<String>> {
    val obj = value.objOrNull ?: return emptyMap()
    fun parse(obj: JsonObject): Map<String, List<String>> = obj.mapNotNull { (key, value) ->
        val array = value.arrayOrNull ?: value.objOrNull?.get("answers").arrayOrNull ?: return@mapNotNull null
        key to array.mapNotNull { it.stringOrNull }
    }.toMap()
    val direct = parse(obj)
    if (direct.isNotEmpty()) return direct
    return obj["answers"].objOrNull?.let(::parse).orEmpty()
}

private fun requestQuestionDetails(
    questions: List<RequestUserInputQuestion>, answers: Map<String, List<String>>,
): List<QuestionDetail> = questions.map { question ->
    val selected = mutableSetOf<String>()
    val other = mutableListOf<String>()
    var note: String? = null
    for (value in answers[question.id].orEmpty()) {
        val trimmed = value.trim()
        val option = question.options.firstOrNull { it.label == trimmed }
        when {
            // An actual option named "user_note: ..." is not metadata.
            option != null -> selected += option.label
            value.startsWith("user_note: ") -> {
                val text = value.removePrefix("user_note: ")
                if (question.inputType == "editor") note = text // Empty documents are answers, too.
                else text.trim().takeIf(String::isNotEmpty)?.let { note = it }
            }
            trimmed.isNotEmpty() -> other += trimmed // Codex history can contain plain free text.
        }
    }
    QuestionDetail(
        header = null, question = question.question, multiple = question.multiple,
        options = question.options.map { QuestionOptionDetail(it.label, it.description, it.label in selected) },
        otherAnswers = other, note = note,
    )
}
